firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Protocols/messages.txt

```diff
@@ -0,0 +1,181 @@
+/* Write function
+Scenario: Message: The write method correctly generates an encoded body from the protobuf message descriptor payload
+
+  GIVEN("A Message object constructed with StateSegment Taggedunion, default message types and a max size of 252 bytes")
+  // assertions carried out in given:
+  // output buffer is empty
+
+    WHEN("The value of the message payload tag is greater than the size of the message descriptor array")
+      THEN("The write method reports invalid type status")
+      THEN("After the write method is called, the output buffer remains unchanged")
+
+    WHEN("The message payload tag is such that it corresponds to Unrecognized message descriptor in descriptor array")
+      THEN("The write method reports invalid type status")
+      THEN("After the write method is called, the output buffer remains unchanged")
+
+    WHEN("The output buffer is too small to hold the generated body")
+      THEN("The write method reports invalid length status")
+      THEN("After the write method is called, the output buffer remains unchanged")
+
+    WHEN("The size of the output buffer is zero")
+      THEN("The write method reports invalid length status")
+
+    // Intent: To show that the write method does not change the internal "type" member of the message class
```

Comment:
I'm surprised, because the first line of the `write` method is: `auto type = static_cast<uint8_t>(payload.tag);`. I must be missing/misunderstanding something here

Reply:
Let's make the `Message::write` method's interface more consistent with the `Datagram::write` and `CRCElement::write` method interfaces. So the contact the method should uphold is that it will update the type member, and the method is not const. This also means that `MessageSender::transform` should take a `const TaggedUnion &payload` (where `TaggedUnion` is now a template parameter for `MessageSender`) instead of `const Message &input_message`; and `BackendSender::transform` should similary take a `const Application::StateSegment &state_segment` instead of a `const BackendMessage &input_message`. Then `Backend::output`, instead of instantiating `BackendMessage message` and calling `synchronizer_.output(message.payload)`, should instantiate an `Application::StateSegment state_segment` and call `synchronizer_.output(state_segment)`.

Reply:
`MessageSender::transform` will need to instantiate its own `Message input_message` object, and then set `input_message.payload = input_payload`, and hopefully we get copy-by-value (we should verify this by checking that the value of `input_payload` hasn't changed after calling `transform`).